internal/exec/stages/disks/partition_dasd.go

```diff
@@ -0,0 +1,298 @@
+		return matches[1], num, err
+	default:
+		return "", 0, ErrBadFdasdOutput
+	}
+}
+
+// parseDasdDiskAndPartInfo parses device specific information needed for converting the sizes into
+// tracks and also gets the information about the partitions. Option example present in:
+// https://www.ibm.com/support/knowledgecenter/linuxonibm/com.ibm.linux.z.lgdd/lgdd_r_fasdusingoptions.html
+func parseDasdDiskAndPartInfo(infostr string) (util.DiskInfo, error) {
+	diskInfo := util.DiskInfo{}
+	dasdDiskInfoRegex := regexp.MustCompile("\\s+([a-z\\ ]+)\\s\\.+\\:\\s+(\\d+)")
+	scanner := bufio.NewScanner(strings.NewReader(infostr))
+	for scanner.Scan() {
+		line := scanner.Text()
+		attr, value, err := parseDasdDiskLine(dasdDiskInfoRegex, line)
+		if err != nil {
+			return diskInfo, err
+		}
+		if value != -1 {
+			switch attr {
+			case "cylinders":
+				diskInfo.Cylinders = value
+				continue
+			case "tracks per cylinder":
+				diskInfo.TracksPerCylinder = value
+				continue
+			case "blocks per track":
+				diskInfo.BlocksPerTrack = value
+				continue
+			case "bytes per block":
+				diskInfo.LogicalSectorSize = value
+				continue
+			default:
+				break
+			}
+		}
+
+		dasdPartInfoRegex := regexp.MustCompile("([a-zA-Z0-9/.-]+)\\s+(\\d+)\\s+(\\d+)\\s+(\\d+)\\s+(\\d+).+")
+		matches := dasdPartInfoRegex.FindStringSubmatch(line)
+		if len(matches) == 6 {
+			partnum, _ := strconv.Atoi(matches[5])
+			startTrack, _ := strconv.Atoi(matches[2])
+			sizeInTracks, _ := strconv.Atoi(matches[4])
+
+			part := util.PartitionInfo{
+				StartTrack:   startTrack,
+				SizeInTracks: sizeInTracks,
+				Number:       partnum,
+			}
+			diskInfo.Partitions = append(diskInfo.Partitions, part)
+		}
+	}
+
+	if scanner.Err() != nil {
+		return util.DiskInfo{}, fmt.Errorf("reading fdasd output failed: %v", scanner.Err())
+	}
+	if diskInfo.Cylinders == 0 || diskInfo.TracksPerCylinder == 0 || diskInfo.BlocksPerTrack == 0 {
+		return util.DiskInfo{}, fmt.Errorf("reading fdasd output failed: could not read disk info - disk may need to be formatted")
+	}
+
+	return diskInfo, nil
+}
+
+// partitionDasdDisk partitions devAlias according to the spec given by dev
+// since the creation of partitions through fdasd invalidates existing partitions,
+// there are some options will are not necessary/ will not make sense for DASD:
+//  - wipePartitionEntry : all partitions will be recreated through fdasd anyway,
+//                         so there is no point to this option.
+//  - matching of partitions to check if it is the same as the original is also
+//    not useful as the partition will be newly recreated anyway.
+func (s stage) partitionDasdDisk(dev types.Disk, devAlias string) error {
+	if dev.WipeTable != nil && *dev.WipeTable {
+		op := dasdfmt.Begin(s.Logger, devAlias)
+		s.Logger.Info("wiping partition table requested on %q", devAlias)
+		op.Commit()
+	} else {
+		s.Logger.Info("Note: Adding any new partitions will NOT preserve existing partitions")
```

Comment:
Unlike sgdisk, there is no way to just create a single partition while preserving other partitions - or putting this another way - there would have to be a  lot of manual mumbo jumbo and calculations to offset the partitions and determine which ones can be preserved and which ones overwritten. sgdisk pretend does that for you  - i did not want to do that manually for fdasd as it would become really messy.  The partitions can be preserved if the user specifies the part number without start or size, then we would use whatever was already present but if not all bets are off.Maybe i should change this to "might not" as it really depends on how the user specifies the partitions. Also, if the user only deletes partitions, the other partitions are preserved. The case where the partitions are not preserved is when the user explicitly specifies certain partitions to be created. Then, the original layout will not be preserved.

Reply:
my initial thought around this was - only create partitions that the user has asked for and do not preserve the existing layout (unless of course the user specifies a partition with a valid number and does not specify a size or start, then  get it from the existing ones). But i wanted the user to also have the ability to just delete certain partitions and still preserve the layout. Where it gets messy is if the user tries to add and delete at the same time and I did not want to make the implementation overly complex by trying to calculate offsets manually and consolidating unused partitions, etc. So whatever the user asks for through the ignition config will be honored, but if there are any partitions not specified in the ignition config and they exist already they will be wiped out.

Yes  - i agree that ignition must not appear to succeed but do a half baked job of it. Maybe we can just reject certain configs outright to prevent this?

Reply:
ok..i changed the logic to cover all scenarios and made sure that we don't provision something the user didn't ask for.